sync/metrics.go

```diff
@@ -11,40 +12,152 @@ import (
+	)
+	if err != nil {
+		return nil, err
 	}
 
-	callback := func(ctx context.Context, observer metric.Observer) error {
-		observer.ObserveFloat64(totalSynced, float64(m.totalSynced.Load()))
-		return nil
+	trustedPeersOutOfSync, err := meter.Int64Counter(
+		"hdr_tr_peers_out_of_sync",
+		metric.WithDescription("trusted peers out of sync"),
+	)
+	if err != nil {
+		return nil, err
 	}
-	_, err = meter.RegisterCallback(callback, totalSynced)
+
+	laggingHeadersStart, err := meter.Int64Counter(
+		"hdr_sync_lagging_hdr_start",
+		metric.WithDescription("lagging header start"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	subjectiveHead, err := meter.Int64ObservableGauge(
+		"hdr_sync_subjective_head",
+		metric.WithDescription("subjective head height"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	blockTime, err := meter.Float64Histogram(
+		"hdr_sync_actual_blockTime_ts",
+		metric.WithDescription("duration between creation of 2 blocks"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &metrics{
+		totalSyncedInst:       totalSynced,
+		syncLoopStarted:       syncLoopStarted,
+		trustedPeersOutOfSync: trustedPeersOutOfSync,
+		laggingHeadersStart:   laggingHeadersStart,
+		blockTime:             blockTime,
+		subjectiveHeadInst:    subjectiveHead,
+		headersThreshold:      headersThreshold,
+	}
+
+	m.syncReg, err = meter.RegisterCallback(m.observeMetrics, m.totalSyncedInst, m.subjectiveHeadInst)
 	if err != nil {
 		return nil, err
 	}
 
 	return m, nil
 }
 
-// recordTotalSynced records the total amount of synced headers.
+func (m *metrics) observeMetrics(ctx context.Context, obs metric.Observer) error {
+	err := m.observeTotalSynced(ctx, obs)
+	if err != nil {
+		return err
+	}
+
+	return m.observeNewHead(ctx, obs)
+}
+
+func (m *metrics) observeTotalSynced(_ context.Context, obs metric.Observer) error {
+	obs.ObserveInt64(m.totalSyncedInst, m.totalSynced.Load())
+	return nil
+}
+
+func (m *metrics) observeNewHead(_ context.Context, obs metric.Observer) error {
+	obs.ObserveInt64(m.subjectiveHeadInst, m.subjectiveHead.Load())
+	return nil
+}
+
 func (m *metrics) recordTotalSynced(totalSynced int) {
+	m.observe(context.Background(), func(_ context.Context) {
+		m.totalSynced.Add(int64(totalSynced))
+	})
+}
+
+func (m *metrics) recordSyncLoopStarted(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.syncLoopStarted.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) recordTrustedPeersOutOfSync(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.trustedPeersOutOfSync.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) observeNewSubjectiveHead(ctx context.Context, height int64, timestamp time.Time) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.subjectiveHead.Store(height)
+
+		if !m.prevHeader.IsZero() {
+			m.blockTime.Record(ctx, timestamp.Sub(m.prevHeader).Seconds())
+		}
+
+		if time.Since(m.headerReceived) > m.headersThreshold {
+			m.laggingHeadersStart.Add(ctx, 1)
+		}
```

Comment:
Instead of this, I would count how many times the getter.Head has to be called during runtime of syncer (it will tell us how efficient header prop is for syncer and if we're doing additional redundant work)

Reply:
IMO it helps to understand the network latency

Reply:
@vgonkivs if the header is lagging, Getter will be called as it does not pass `isRecent` check. Let's not duplicate logic.

Reply:
+1. I think this should be a separate observe method on metrics struct that is called exactly when we detect lagg in Syncer. This will avoid the need to duplicate logic, as Rene mentioned, as well avoid passing headersThreshold to metrics.

Reply:
okay, so there are two possible places where we can have lagging head:
* inside syncer.Head;
* inside subjectiveHead;

Am I understanding you correctly, that you proposed to add one more metric to distinguish the places where it happened?(E.g for sync.Head we can have syncHeadLagging, for subjHead - subkHeadLagging)

@Wondertan